File: app/src/main/java/ch/threema/app/preference/SettingsAboutFragment.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2016-2021 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.preference;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.view.View;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceScreen;
import ch.threema.app.BuildConfig;
import ch.threema.app.BuildFlavor;
import ch.threema.app.R;
import ch.threema.app.ThreemaApplication;
import ch.threema.app.activities.AboutActivity;
import ch.threema.app.activities.DownloadApkActivity;
import ch.threema.app.activities.LicenseActivity;
import ch.threema.app.activities.PrivacyPolicyActivity;
import ch.threema.app.dialogs.GenericProgressDialog;
import ch.threema.app.dialogs.SimpleStringAlertDialog;
import ch.threema.app.exceptions.FileSystemNotPresentException;
import ch.threema.app.managers.ServiceManager;
import ch.threema.app.services.PreferenceService;
import ch.threema.app.services.license.LicenseService;
import ch.threema.app.services.license.LicenseServiceSerial;
import ch.threema.app.utils.AppRestrictionUtil;
import ch.threema.app.utils.ConfigUtils;
import ch.threema.app.utils.DialogUtil;
import ch.threema.app.utils.IntentDataUtil;
import ch.threema.app.utils.TestUtil;


public class SettingsAboutFragment extends ThreemaPreferenceFragment {
	private static final Logger logger = LoggerFactory.getLogger(SettingsAboutFragment.class);

	private static final int ABOUT_REQUIRED_CLICKS = 10;
	private static final String DIALOG_TAG_CHECK_UPDATE = "checkup";

	private String updateUrl, updateMessage;
	private int aboutCounter;

	private PreferenceService preferenceService;
	private LicenseService licenseService;

	@Override
	public void onCreatePreferencesFix(@Nullable Bundle savedInstanceState, String rootKey) {
		if (!requiredInstances()) {
			return;
		}

		addPreferencesFromResource(R.xml.preference_about);

		PreferenceScreen preferenceScreen = (PreferenceScreen) findPreference("pref_key_about");
		PreferenceCategory aboutCategory = (PreferenceCategory) findPreference("pref_key_about_header");

		Preference aboutPreference = findPreference(getResources().getString(R.string.preferences__about));
		Preference workLicensePreference = findPreference(getResources().getString(R.string.preferences__work_license_name));
		Preference licensePreference = findPreference(getResources().getString(R.string.preferences__licenses));
		Preference privacyPolicyPreference = findPreference(getResources().getString(R.string.preferences__privacy_policy));
		Preference checkUpdatePreference = findPreference(getResources().getString(R.string.preferences__check_updates));
		Preference troubleShootingPreference = findPreference(getResources().getString(R.string.preferences__troubleshooting));
		Preference deviceInfoPreference = findPreference(getResources().getString(R.string.preferences__device_info));
		Preference translatorsPreference = findPreference(getResources().getString(R.string.preferences__translators));

		aboutPreference.setTitle(getString(R.string.threema_version) + " " + getVersionString());
		aboutPreference.setSummary(R.string.about_copyright);
		aboutPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				if (aboutCounter == ABOUT_REQUIRED_CLICKS) {
					aboutCounter++;
					final Intent intent = new Intent(getActivity().getApplicationContext(), AboutActivity.class);
					intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_TASK_ON_HOME);
					startActivity(intent);
					final Activity activity = getActivity();
					if (activity != null) {
						activity.finish();
					}
					return true;
				} else {
					aboutCounter++;
					return false;
				}
			}
		});

		licensePreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				startActivity(new Intent(getActivity().getApplicationContext(), LicenseActivity.class));
				return true;
			}
		});

		privacyPolicyPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				startActivity(new Intent(getActivity().getApplicationContext(), PrivacyPolicyActivity.class));
				return true;
			}
		});

		if (ConfigUtils.isSerialLicensed() && !ConfigUtils.isWorkBuild()) {
			checkUpdatePreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
				@Override
				public boolean onPreferenceClick(Preference preference) {
					checkForUpdates((LicenseServiceSerial) licenseService);
					return true;
				}
			});
		} else {
			aboutCategory.removePreference(checkUpdatePreference);
		}

		if (ConfigUtils.isWorkBuild()) {
			workLicensePreference.setSummary(preferenceService.getLicenseUsername());
			preferenceScreen.removePreference(findPreference("pref_key_feedback_header"));

			if (ConfigUtils.isWorkRestricted()) {
				Boolean readonly = AppRestrictionUtil.getBooleanRestriction(getString(R.string.restriction__readonly_profile));

				if (readonly != null && readonly) {
					aboutCategory.removePreference(workLicensePreference);
				}
			}
		} else {
			aboutCategory.removePreference(workLicensePreference);
		}

		if (deviceInfoPreference != null) {
			if (Build.MANUFACTURER != null) {
				deviceInfoPreference.setTitle(Build.MANUFACTURER + " " + Build.MODEL);
			}
			deviceInfoPreference.setSummary(Build.FINGERPRINT);
		}

		if (troubleShootingPreference != null) {
			troubleShootingPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
				@Override
				public boolean onPreferenceClick(Preference preference) {
					final Header header = new Header();
					header.fragment = "ch.threema.app.preference.SettingsTroubleshootingFragment";
					header.iconRes = R.drawable.ic_bug_outline;
					header.titleRes = R.string.prefs_troubleshooting;

					try {
						((SettingsActivity) getActivity()).switchToHeader(header);
					} catch (Exception ignore) {}
					return true;
				}
			});
		}

		translatorsPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				SimpleStringAlertDialog.newInstance(R.string.translators, getString(R.string.translators_thanks, getString(R.string.translators_list))).show(getFragmentManager(), "tt");
				return true;
			}
		});

	}

	final protected boolean requiredInstances() {
		if (!this.checkInstances()) {
			this.instantiate();
		}
		return this.checkInstances();
	}

	protected boolean checkInstances() {
		return TestUtil.required(
				this.preferenceService,
				this.licenseService
		);
	}

	protected void instantiate() {
		ServiceManager serviceManager = ThreemaApplication.getServiceManager();
		if (serviceManager != null) {
			this.preferenceService = serviceManager.getPreferenceService();
			try {
				this.licenseService = serviceManager.getLicenseService();
			} catch (FileSystemNotPresentException e) {
				logger.error("Exception", e);
			}
		}
	}

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		preferenceFragmentCallbackInterface.setToolbarTitle(R.string.menu_about);
		super.onViewCreated(view, savedInstanceState);
	}

	private String getVersionString() {
		final Context context = getContext();
		final StringBuilder version = new StringBuilder();
		version.append(ConfigUtils.getFullAppVersion(context));
		version.append(" Build ").append(ConfigUtils.getBuildNumber(context));
		version.append(" ").append(BuildFlavor.getName());
		if (BuildConfig.DEBUG) {
			version.append(" Commit ").append(BuildConfig.GIT_HASH);
		}
		return version.toString();
	}

	@SuppressLint("StaticFieldLeak")
	private void checkForUpdates(final @NonNull LicenseServiceSerial licenseServiceSerial) {
		if (BuildConfig.FLAVOR.equals("store_threema")) {
				new AsyncTask<Void, Void, String>() {
					@Override
					protected void onPreExecute() {
						GenericProgressDialog.newInstance(R.string.check_updates, R.string.please_wait).show(getFragmentManager(), DIALOG_TAG_CHECK_UPDATE);
					}

					@Override
					protected String doInBackground(Void... voids) {
						try {
							// Validate license and check for updates
							licenseServiceSerial.validate(false);

							// If an update is available, then `updateUrl` and `updateMessage` will
							// be set to a non-null value.
							updateUrl = licenseServiceSerial.getUpdateUrl();
							updateMessage = licenseServiceSerial.getUpdateMessage();
							if (TestUtil.empty(updateUrl, updateMessage)) {
								// No update available...
								return getString(R.string.no_update_available);
							}

							// Update available!
							return null;
						} catch (final Exception x) {
							return String.format(getString(R.string.an_error_occurred_more), x.getLocalizedMessage());
						}
					}

					@Override
					protected void onPostExecute(String error) {
						DialogUtil.dismissDialog(getFragmentManager(), DIALOG_TAG_CHECK_UPDATE, true);
						if (error != null) {
							SimpleStringAlertDialog.newInstance(R.string.check_updates, error).show(getFragmentManager(), "nu");
						} else {
							Intent dialogIntent = IntentDataUtil.createActionIntentUpdateAvailable(updateMessage, updateUrl);
							dialogIntent.putExtra(DownloadApkActivity.EXTRA_FORCE_UPDATE_DIALOG, true);
							dialogIntent.setClass(getContext(), DownloadApkActivity.class);
							startActivity(dialogIntent);
						}
					}
				}.execute();
		}
	}
}

File: app/src/main/java/ch/threema/app/preference/SettingsAboutFragment.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2022-2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.preference

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Intent
import android.os.Build
import androidx.core.net.toUri
import androidx.lifecycle.lifecycleScope
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import ch.threema.android.showToast
import ch.threema.app.BuildConfig
import ch.threema.app.BuildFlavor
import ch.threema.app.R
import ch.threema.app.activities.DownloadApkActivity
import ch.threema.app.dialogs.GenericProgressDialog
import ch.threema.app.dialogs.SimpleStringAlertDialog
import ch.threema.app.preference.service.PreferenceService
import ch.threema.app.restrictions.AppRestrictionUtil
import ch.threema.app.services.license.LicenseService
import ch.threema.app.services.license.LicenseServiceSerial
import ch.threema.app.utils.ConfigUtils
import ch.threema.app.utils.DialogUtil
import ch.threema.app.utils.DispatcherProvider
import ch.threema.app.utils.IntentDataUtil
import ch.threema.app.utils.logScreenVisibility
import ch.threema.app.webviews.EulaActivity
import ch.threema.app.webviews.LicenseActivity
import ch.threema.app.webviews.PrivacyPolicyActivity
import ch.threema.app.webviews.TermsOfServiceActivity
import ch.threema.base.utils.getThreemaLogger
import ch.threema.common.takeUnlessEmpty
import ch.threema.localcrypto.MasterKeyManager
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.android.ext.android.inject

private val logger = getThreemaLogger("SettingsAboutFragment")

@Suppress("unused")
class SettingsAboutFragment : ThreemaPreferenceFragment() {
    init {
        logScreenVisibility(logger)
    }

    private var aboutCounter = 0

    private val preferenceService: PreferenceService by inject()
    private val licenseService: LicenseService<*> by inject()
    private val masterKeyManager: MasterKeyManager by inject()
    private val dispatcherProvider: DispatcherProvider by inject()

    override fun initializePreferences() {
        initLicensePref()
        initPrivacyPolicyPref()
        initTermsOfServicePref()
        initEndUserLicensePref()
        initVersionSection()
        initSelfUpdatePref()
        initServerConfigSection()
    }

    override fun getPreferenceTitleResource(): Int = R.string.menu_about

    override fun getPreferenceResource(): Int = R.xml.preference_about

    private fun initLicensePref() {
        val licensePreference = getPref<Preference>(R.string.preferences__licenses)
        licensePreference.onClick {
            startActivity(LicenseActivity.createIntent(requireContext()))
        }
    }

    private fun initPrivacyPolicyPref() {
        val privacyPolicyPreference = getPref<Preference>(R.string.preferences__privacy_policy)
        if (ConfigUtils.isOnPremBuild() && !ConfigUtils.isDemoOPServer(preferenceService)) {
            privacyPolicyPreference.isVisible = false
        } else {
            privacyPolicyPreference.onClick {
                startActivity(PrivacyPolicyActivity.createIntent(requireContext()))
            }
        }
    }

    private fun initTermsOfServicePref() {
        val licensePreference = getPref<Preference>(R.string.preferences__terms_of_service)
        if (BuildFlavor.current.licenseType == BuildFlavor.LicenseType.ONPREM) {
            licensePreference.isVisible = false
        } else {
            licensePreference.onClick {
                startActivity(TermsOfServiceActivity.createIntent(requireContext()))
            }
        }
    }

    private fun initEndUserLicensePref() {
        val licensePreference = getPref<Preference>(R.string.preferences__eula)
        if (BuildFlavor.current.licenseType == BuildFlavor.LicenseType.GOOGLE) {
            licensePreference.onClick {
                startActivity(EulaActivity.createIntent(requireContext()))
            }
        } else {
            licensePreference.isVisible = false
        }
    }

    private fun initVersionSection() {
        initVersionPref()
        initVersionCodePref()
        initDeviceInfoPref()
        initDebugInfoPrefs()
    }

    private fun initVersionPref() {
        val versionPreference = getPref<Preference>(R.string.preferences__version)
        versionPreference.title = getVersionNameWithBuildNumber()
        versionPreference.setSummary(R.string.about_copyright)
        versionPreference.onClick {
            aboutCounter++
            if (aboutCounter == 10) {
                onSecretUnlocked()
            }
        }
    }

    private fun onSecretUnlocked() {
        if (ConfigUtils.isDevBuild() && !preferenceService.showDeveloperMenu()) {
            preferenceService.setShowDeveloperMenu(true)
            showToast("Developer settings unlocked")
        }

        val installerPackage = ConfigUtils.getInstallerPackageName(requireContext())
        getPref<Preference>(R.string.preferences__installer_package).summary = installerPackage ?: "none"
        getPref<Preference>(R.string.preferences__hidden_info_section).isVisible = true
    }

    private fun initVersionCodePref() {
        getPref<Preference>(R.string.preferences__version_code).apply {
            title = getString(R.string.threema_version_code)
            summary = BuildConfig.DEFAULT_VERSION_CODE.toString()
        }
    }

    private fun initDeviceInfoPref() {
        getPrefOrNull<Preference>(R.string.preferences__device_info)?.apply {
            if (Build.MANUFACTURER != null) {
                title = Build.MANUFACTURER + " " + Build.MODEL
            }
            summary = Build.FINGERPRINT
        }
    }

    private fun initDebugInfoPrefs() {
        if (!BuildConfig.DEBUG) {
            return
        }
        BuildConfig.GIT_HASH.takeUnlessEmpty().let { hash ->
            getPref<Preference>(R.string.preferences__git_commit).apply {
                isVisible = true
                summary = hash
            }
        }
        BuildConfig.GIT_BRANCH.takeUnlessEmpty().let { branch ->
            getPref<Preference>(R.string.preferences__git_branch).apply {
                isVisible = true
                summary = branch
            }
        }
    }

    private fun initSelfUpdatePref() {
        val checkUpdatePreference = getPref<Preference>(R.string.preferences__check_updates)

        if (BuildFlavor.current.maySelfUpdate) {
            checkUpdatePreference.onClick {
                checkForUpdates(licenseService as LicenseServiceSerial)
            }
        } else {
            val aboutCategory = getPref<PreferenceCategory>("pref_key_about_header")
            aboutCategory.removePreference(checkUpdatePreference)
        }
    }

    private fun initServerConfigSection() {
        val shouldShowServer = ConfigUtils.isOnPremBuild()
        val shouldShowUsername = shouldShowUsername()
        val shouldShowRemoteSecretActivated = masterKeyManager.isProtectedWithRemoteSecret() == true

        if (!shouldShowServer && !shouldShowUsername && !shouldShowRemoteSecretActivated) {
            getPref<Preference>(R.string.preferences__server_config).isVisible = false
            return
        }

        getPref<Preference>(R.string.preferences__work_license_name)
            .let { workLicensePreference ->
                if (shouldShowUsername) {
                    workLicensePreference.summary = preferenceService.licenseUsername
                }
                workLicensePreference.isVisible = shouldShowUsername
            }

        getPref<Preference>(R.string.preferences__onprem_server)
            .let { serverConfigPreference ->
                if (shouldShowServer) {
                    serverConfigPreference.summary = getServerInfo()
                }
                serverConfigPreference.isVisible = shouldShowServer
            }

        getPref<Preference>(R.string.preferences__remote_secret_activated)
            .let { remoteSecretPreference ->
                if (shouldShowRemoteSecretActivated) {
                    remoteSecretPreference.onClick(::onRemoteSecretClicked)
                } else {
                    remoteSecretPreference.isVisible = false
                }
            }
    }

    private fun shouldShowUsername(): Boolean {
        return when {
            !ConfigUtils.isWorkBuild() -> false
            ConfigUtils.isWorkRestricted() -> {
                AppRestrictionUtil.getBooleanRestriction(getString(R.string.restriction__readonly_profile)) != true
            }
            else -> true
        }
    }

    private fun getServerInfo(): String {
        return preferenceService.onPremServer
            ?.toUri()
            ?.authority
            ?: "?"
    }

    private fun getVersionNameWithBuildNumber() = buildString {
        appendVersionName()
        appendBuildNumber()
        appendBuildFlavor()
    }

    private fun StringBuilder.appendVersionName() {
        append(getString(R.string.threema_version))
        append(" ")
        append(BuildConfig.VERSION_NAME)
    }

    private fun StringBuilder.appendBuildNumber() {
        append(" Build ")
        append(ConfigUtils.getBuildNumber(context))
    }

    private fun StringBuilder.appendBuildFlavor() {
        append(" ")
        append(BuildFlavor.current.fullDisplayName)
    }

    private fun onRemoteSecretClicked() {
        try {
            val learnMoreUrl = getString(R.string.remote_secret_learn_more_url).toUri()
            startActivity(Intent(Intent.ACTION_VIEW, learnMoreUrl))
        } catch (e: ActivityNotFoundException) {
            logger.warn("No activity found to open Remote Secret learn more URL")
        }
    }

    @Suppress("DEPRECATION")
    @SuppressLint("StaticFieldLeak")
    private fun checkForUpdates(licenseServiceSerial: LicenseServiceSerial) {
        if (!BuildFlavor.current.maySelfUpdate) {
            logger.warn("Called checkForUpdate in a build variant without self-updating")
            return
        }

        lifecycleScope.launch {
            activity?.let { activity ->
                GenericProgressDialog.newInstance(R.string.check_updates, R.string.please_wait)
                    .show(activity.supportFragmentManager, DIALOG_TAG_CHECK_UPDATE)
            }
            val result = withContext(dispatcherProvider.worker) {
                try {
                    licenseServiceSerial.validate(false)
                    val updateUrl = licenseServiceSerial.updateUrl
                    if (!updateUrl.isNullOrEmpty()) {
                        Result.UpdateUrl(updateUrl)
                    } else {
                        Result.Error(getString(R.string.no_update_available))
                    }
                } catch (e: Exception) {
                    logger.error("Failed to get update url")
                    Result.Error(getString(R.string.an_error_occurred_more, e.localizedMessage))
                }
            }

            DialogUtil.dismissDialog(parentFragmentManager, DIALOG_TAG_CHECK_UPDATE, true)
            when (result) {
                is Result.UpdateUrl -> {
                    val updateMessage = getString(R.string.update_available_message)
                    val dialogIntent =
                        IntentDataUtil.createActionIntentUpdateAvailable(updateMessage, result.url)
                            .putExtra(DownloadApkActivity.EXTRA_FORCE_UPDATE_DIALOG, true)
                            .setClass(requireContext(), DownloadApkActivity::class.java)
                    startActivity(dialogIntent)
                }
                is Result.Error -> {
                    SimpleStringAlertDialog.newInstance(R.string.check_updates, result.error)
                        .show(parentFragmentManager, "nu")
                }
            }
        }
    }

    private sealed class Result {
        data class UpdateUrl(val url: String) : Result()
        data class Error(val error: String) : Result()
    }

    companion object {
        private const val DIALOG_TAG_CHECK_UPDATE = "checkup"
    }
}
